test/integration/utils.go

```diff
@@ -0,0 +1,499 @@
+	case string(corev1.Linux):
+		yamlPath = cnsLinuxDaemonSetPath
+	case string(corev1.Windows):
+		yamlPath = cnsWindowsDaemonsetPath
+	default:
+		return appsv1.DaemonSet{}, ErrUnsupportedOS
+	}
+
+	cns, err := k8sutils.MustParseDaemonSet(yamlPath)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrapf(err, "failed to parse daemonset %s", yamlPath)
+	}
+	return cns, nil
+}
+
+func loadCNSDaemonset(ctx context.Context, clientset *kubernetes.Clientset, cnsVersion, cniDropgzVersion, nodeOS string) (appsv1.DaemonSet, error) {
+	cns, err := ParseCNSDaemonset(nodeOS)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrapf(err, "failed to parse daemonset for %s", nodeOS)
+	}
+
+	labelSelector := ""
+	switch nodeOS {
+	case string(corev1.Linux):
+		labelSelector = cnsLinuxLabelSelector
+	case string(corev1.Windows):
+		labelSelector = cnsWindowsLabelSelector
+	default:
+		return appsv1.DaemonSet{}, ErrUnsupportedOS
+	}
+
+	image, _ := k8sutils.ParseImageString(cns.Spec.Template.Spec.Containers[0].Image)
+	cns.Spec.Template.Spec.Containers[0].Image = k8sutils.GetImageString(image, cnsVersion)
+
+	log.Printf("Checking environment scenario")
+	cns = loadDropgzImage(cns, cniDropgzVersion)
+	cns, err = loadAzureVnet(ctx, clientset, cns)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to load azure vnet")
+	}
+	cns, err = loadAzilium(ctx, clientset, cns)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to load azilium")
+	}
+	cns, err = loadCiliumOverlay(ctx, clientset, cns)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to load cilium overlay")
+	}
+	cns, err = loadAzureCNIOverlay(ctx, clientset, cns, nodeOS)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to load azure cni overlay")
+	}
+	cns, err = loadDualstackOverlay(ctx, clientset, cns)
+	if err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to load dualstack overlay")
+	}
+
+	cnsDaemonsetClient := clientset.AppsV1().DaemonSets(cns.Namespace)
+
+	log.Printf("Installing CNS with image %s", cns.Spec.Template.Spec.Containers[0].Image)
+
+	// setup common RBAC, ClusteerRole, ClusterRoleBinding, ServiceAccount
+	if _, err := k8sutils.MustSetUpClusterRBAC(ctx, clientset, cnsClusterRolePath, cnsClusterRoleBindingPath, cnsServiceAccountPath); err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to setup common RBAC, ClusteerRole, ClusterRoleBinding and ServiceAccount")
+	}
+
+	// setup RBAC, Role, RoleBinding
+	if err := k8sutils.MustSetUpRBAC(ctx, clientset, cnsRolePath, cnsRoleBindingPath); err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to setup RBAC, Role and RoleBinding")
+	}
+
+	if err := k8sutils.MustCreateDaemonset(ctx, cnsDaemonsetClient, cns); err != nil {
+		return appsv1.DaemonSet{}, errors.Wrap(err, "failed to create daemonset")
+	}
+
+	if err := k8sutils.WaitForPodsRunning(ctx, clientset, cns.Namespace, labelSelector); err != nil {
```

Comment:
`k8sutils.WaitForPodDaemonset` is more resilient and has the same functionality as `k8sutils.WaitForPodsRunning`. If the daemonset already exists then it won't continue until all pods have been restarted otherwise it will proceed as intended.